File: rust_euler/20-40/p021.rs
/* Project Euler - Problem 21
  Evaluate the sum of all the amicable numbers under 10000.
  
  p021.rs
  rustc p021.rs
  ./p021
  
  notes:
    amicable: d(a) = b && d(b) = a, a!=b; 
              where d(n) is the sum of proper divisors (n%m = 0, m < n)
              
*/

fn main() {
    let num = 10000;
    println!("num: {}", num);
    
    if true {
        println!("sum_prop_div(284): {}", sum_prop_div(284))
    }
    
    let result = p021(num);
    println!("result: {}", result)
}

fn p021(num: u32) -> u32 {
    let mut sum_am = 0;
    for n in 2..num {
        let n_am = sum_prop_div(n);
        if n == sum_prop_div(n_am) && n != n_am {
            sum_am += n;
        }
    }
    return sum_am;
}

fn sum_prop_div (num: u32) -> u32 {
    let mut sum = 1;
    let sqrt_num = (num as f64).sqrt() as u32 + 1;
    for i in 2..sqrt_num {
        if num%i == 0 {
            if i == num/i {
                sum += i;
            } else {
                sum += i + num/i;
            }
        }
    }
    return sum;
}

File: julia_euler/20-40/p021.jl
# Project Euler - Problem 021
# Evaluate the sum of all the amicable numbers under 10000.
#
# julia p021.jl
#
# notes:
#   amicable: d(a) = b && d(b) = a, a!=b; 
#             where d(n) is the sum of proper divisors (n%m = 0, m < n)

function sum_prop_div(num)
    sum = 1
    for i = 1:floor(sqrt(num))
        if num%i == 0
            if i == num/i
                sum += i
            else
                sum += i + num/i
            end
        end
    end
    return sum
end

function ammicable(n)
    n_am = sum_prop_div(n)
    return n == sum_prop_div(n_am) && n != n_am ? n : 0
end

lim = 9999
println("ammicable(240): ", ammicable(240))
println(sum([ammicable(n) for n in 2:lim]))
